src/storage/segcache/src/segcache.rs

```diff
@@ -0,0 +1,288 @@
+    /// increasing the item frequency - useful for combined operations that
+    /// check for presence - eg replace is a get + set
+    /// ```
+    /// use segcache::{CoarseDuration, Policy, SegCache};
+    ///
+    /// let mut cache = SegCache::builder().build();
+    /// assert!(cache.get_no_freq_incr(b"coffee").is_none());
+    /// ```
+    pub fn get_no_freq_incr(&mut self, key: &[u8]) -> Option<Item> {
+        self.hashtable.get_no_freq_incr(key, &mut self.segments)
+    }
+
+    /// Insert a new item into the cache. May return an error indicating that
+    /// the insert was not successful.
+    /// ```
+    /// use segcache::{CoarseDuration, Policy, SegCache};
+    ///
+    /// let mut cache = SegCache::builder().build();
+    /// assert!(cache.get(b"drink").is_none());
+    ///
+    /// cache.insert(b"drink", b"coffee", None, CoarseDuration::ZERO);
+    /// let item = cache.get(b"drink").expect("didn't get item back");
+    /// assert_eq!(item.value(), b"coffee");
+    ///
+    /// cache.insert(b"drink", b"whisky", None, CoarseDuration::ZERO);
+    /// let item = cache.get(b"drink").expect("didn't get item back");
+    /// assert_eq!(item.value(), b"whisky");
+    /// ```
+    pub fn insert<'a>(
+        &mut self,
+        key: &'a [u8],
+        value: &[u8],
+        optional: Option<&[u8]>,
+        ttl: CoarseDuration,
+    ) -> Result<(), SegCacheError<'a>> {
+        // default optional data is empty
+        let optional = optional.unwrap_or(&[]);
+
+        // calculate size for item
+        let size = (((ITEM_HDR_SIZE + key.len() + value.len() + optional.len()) >> 3) + 1) << 3;
+
+        // try to get a `ReservedItem`
+        let mut retries = RESERVE_RETRIES;
+        let reserved;
+        loop {
+            match self
+                .ttl_buckets
+                .get_mut_bucket(ttl)
+                .reserve(size, &mut self.segments)
+            {
+                Ok(mut reserved_item) => {
+                    reserved_item.define(key, value, optional);
+                    reserved = reserved_item;
+                    break;
+                }
+                Err(TtlBucketsError::ItemOversized { size }) => {
+                    return Err(SegCacheError::ItemOversized { size, key });
+                }
+                Err(TtlBucketsError::NoFreeSegments) => {
+                    if self
+                        .segments
+                        .evict(&mut self.ttl_buckets, &mut self.hashtable)
+                        .is_err()
+                    {
+                        retries -= 1;
+                    } else {
+                        continue;
+                    }
+                }
+            }
+            if retries == 0 {
+                return Err(SegCacheError::NoFreeSegments);
+            }
+            retries -= 1;
+        }
+
+        // insert into the hashtable, or roll-back by removing the item
+        // TODO(bmartin): we can probably roll-back the offset and re-use the
```

Comment:
I wonder what our policy with overfull HT should be eventually. My intuition is that:
- if eviction is not permitted, we should roll back and fail the insertion so we don't violate our promised SLO;
- if eviction is permitted, we should evict an existing item, which is older than the current one, since age correlate strongly (though may not be the No.1 indicator) of usefulness of the item.

Reply:
Yeah, I think we can do something better here for the overfull HT case, there's no reason (in the non-concurrent implementation) that we can't roll-back the write offset in the segment to avoid wasting space.

We'd need to evaluate the impact of looking at item ages across all item entries in the corresponding hash bucket(s) and how that compares w/ doing random item within the same hash bucket(s) - technically we could do a similar set of eviction strategies as we have for segments, LRU, CTE, Random would all probably be worth exploring?

Reply:
we can just evict something from the same hash bucket by unlinking it, right? Since the offset reflects the order items were inserted/linked.